src/dotnet/ShellShim/ShellShimRepositoryFactory.cs

```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.DotNet.Configurer;
+using Microsoft.Extensions.EnvironmentAbstractions;
+
+namespace Microsoft.DotNet.ShellShim
+{
+    internal class ShellShimRepositoryFactory : IShellShimRepositoryFactory
+    {
+        public IShellShimRepository CreateShellShimRepository(DirectoryPath? nonGlobalLocation = null)
+        {
+            return new ShellShimRepository(GetShimLocation(nonGlobalLocation));
```

Comment:
I'd prefer this to be `nonGlobalLocation ?? GetGlobalShimLocation()` rather than passing it through to a function whose job it is to return it back if not null.

Reply:
no that function need to call CliFolderPathCalculator,

var cliFolderPathCalculator = new CliFolderPathCalculator();
packageLocation = new DirectoryPath(cliFolderPathCalculator.ToolsShimPath);

Reply:
ah, i see, nwm, no strong opinion on that

Reply:
Thinking about this more, though, this is a lot of abstraction for what really boils down to getting the global shim location.  I'm not sure the factory pattern is useful here because all this is doing is calculating the global shim location and constructing only one possible type to give back (i.e. it's simply abstracting away a constructor call).  This is only really saving the users of this factory from doing something like `new ShellShimRepository(myLocation ?? new DirectoryPath(new CliFolderPathCalculator().ToolsShimPath)`.

Reply:
Aside: why isn't `CliFolderPathCalculator` static? It has no fields :/

Reply:
See my description top comment. Once you `new ShellShimRepository` the location is fixed to .dotnet, and we cannot swap implementation anymore

Reply:
Ok, now I'm on the same page.  I retract my concern over the complexity (the original comment on this thread, though, I leave in place).